### DISCUSSION TITLE
How to clean up if a an action fails?

### DISCUSSION THREAD

**Discussion by AlejandroAkbal** [lorisleiva/laravel-actions#186] [Q&A] [answered]
I've recreated the steps defined here:
https://laravel.com/docs/9.x/queues#cleaning-up-after-failed-jobs

But the `failed` method does not seem to be triggered when an exception is thrown

**Comment by Wulfheart** [comment]
Do you have a reproduction repository?

**Reply by AlejandroAkbal** [reply]
No, sorry

I'm not near a computer right now, but a simple action that throws an exception and has a "failed" public method as per the jobs docs should suffice

**Comment by lorisleiva** [comment]
You're pointing to the Laravel documentation of a job which differs from an Action being run as a Job.

An Action as defined by this package is not a Job as defined by Laravel. Please see https://laravelactions.com/2.x/how-does-it-work.html#decorating-your-actions.

The failed hook is not currently supported. Please see `AsJob` references here: https://laravelactions.com/2.x/as-job.html.

**Reply by AlejandroAkbal** [reply]
So no way to clean up after jobs then, right?

**Reply by AlejandroAkbal** [reply]
I'll be wrapping everything in a try-catch then

**Comment by jameswagoner** [comment, marked as answer]
I am actually doing something similar to a clean up and works well.

Just type hint `JobDecorator` in your `asJob()` method. Then you can add `jobFailed()` method to cleanup.

```php

public function handle(Team $team, bool $fullReport = false): void
{
    // Prepare report and send it to all $team->users.
}

public function asJob(JobDecorator $job, Team $team)
{
    if ($team->someCondition) {
        $job->fail('Some condition failed this job');

        return;
    }

    $this->handle($team, true);
}

public function jobFailed(Throwable $e): void
{
    // Cleanup
}
